### PR TITLE
Implements IcePop and creates a unified interface for off-policy correction

### PR DESCRIPTION
**Pull request by finbarrtimbers**

### CODE REVIEW

File: open_instruct/grpo_utils.py [hunk 1]
```diff
@@ -102,7 +102,15 @@ class GRPOExperimentConfig(
...
-    """The maximum cap for truncated importance sampling ratio (0 means disabled)"""
+    """The maximum cap for truncated importance sampling ratio (0 means disabled)."""
+    use_icepop: bool = False
+    """If True, replace truncated importance sampling with the IcePop M(ρ; 1/β, β)
+    operator from https://arxiv.org/abs/2510.18855: tokens whose train/infer
+    mismatch ratio ρ = π^train_old / π^infer_old falls in [1/β, β] are reweighted
+    by ρ (an IS correction); tokens outside that band have their per-token
+    policy loss zeroed out."""
+    icepop_beta: float = 5.0
+    """IcePop bound β (paper default: 5.0); the in-range band is [1/β, β]."""
```
Comment (hamishivi) [inline open_instruct/grpo_utils.py]: Would it be better to have icepop_upper and icepop_lower, in case you don't want the strict 1/b, b range? I think in GLM-5 paper they just do lower/higher (https://arxiv.org/pdf/2602.15763).

File: open_instruct/grpo_utils.py [hunk 2]
```diff
@@ -320,21 +330,71 @@ def compute_vllm_local_debug_metrics(
...
     if cap <= 0:
-        return None, None
-    unclamped = torch.ones_like(old_logprob)
-    logprob_diff = old_logprob - vllm_logprobs
-    logprob_diff = torch.where(response_mask, logprob_diff.clamp(-10.0, 10.0), torch.zeros_like(logprob_diff))
-    unclamped = torch.where(response_mask, torch.exp(logprob_diff), unclamped)
-    clamped = torch.clamp(unclamped, max=cap)
-    return clamped, unclamped
+        return RhoCorrection(weights=None, metrics={}, histogram_metrics=rho_hist)
+    clamped = torch.clamp(rho, max=cap)
```
Comment (hamishivi) [inline open_instruct/grpo_utils.py]: So for icepop, we apply the TIS weights AND pop/mask the ones lying too far out?

File: open_instruct/grpo_utils.py [hunk 3]
```diff
@@ -320,21 +330,71 @@ def compute_vllm_local_debug_metrics(
...
+    (response tokens only); these bypass the scalar reduction and are
+    concatenated across micro-batches and logged as wandb histograms.
     """
+
+    weights: torch.Tensor | None
+    metrics: dict[str, torch.Tensor]
+    histogram_metrics: dict[str, torch.Tensor] = field(default_factory=dict)
+
+
+def compute_rho_correction(
```
Comment (hamishivi) [inline open_instruct/grpo_utils.py:391]: It might be nice to also have a sequence level version or flag, potentially...?

**Review by hamishivi — CHANGES_REQUESTED** [review]
Generally seems good. just some minor comments.

File: open_instruct/grpo_utils.py (hunk 1 above)
Comment (finbarrtimbers) [reply open_instruct/grpo_utils.py]: yeah good point, done.

File: open_instruct/grpo_utils.py (hunk 2 above)
Comment (finbarrtimbers) [reply open_instruct/grpo_utils.py]: Correct. This is what Kimi and verl do (but not what the original paper does).

File: open_instruct/grpo_utils.py (hunk 3 above)
Comment (finbarrtimbers) [reply open_instruct/grpo_utils.py:391]: so we'd compute it as the average over the sequence? Or how would you imagine that working?

Comment (hamishivi) [reply open_instruct/grpo_utils.py:391]: yeah. like deepseek v3.2
image

File: scripts/train/qwen/qwen3_4b_dapo_math.sh
Comment (hamishivi) [inline scripts/train/qwen/qwen3_4b_dapo_math.sh:1]: can we revert these changed unless you want to include this as the new baseline?

File: scripts/submit_eval_jobs.py
Comment (hamishivi) [inline scripts/submit_eval_jobs.py:1]: Can we put this in a sep pr? At least I'd like it if we had the old submit eval script still around.

File: open_instruct/grpo_utils.py [hunk 4]
```diff
@@ -102,7 +102,22 @@ class GRPOExperimentConfig(
     clip_higher: float = 0.272
     """the higher clip range. Sometimes we want this to be higher, see DAPO (https://arxiv.org/abs/2503.14476)"""
     truncated_importance_sampling_ratio_cap: float = 2.0
-    """The maximum cap for truncated importance sampling ratio (0 means disabled)"""
+    """The maximum cap for truncated importance sampling ratio (0 means disabled)."""
+    use_icepop: bool = False
```
Comment (hamishivi) [inline open_instruct/grpo_utils.py]: Can we allow stacking TIS and icepop? This allows us to e.g. recreate what periodic is doing. (source: https://x.com/agarwl_/status/2049263264249368906?s=20)

Image

Comment (hamishivi) [reply open_instruct/grpo_utils.py]: So then we have a mask + a ratio.

**Review by hamishivi — APPROVED** [review]
LGTM! One thing is we could make the naming more generic for users, instead of using 'icepop'.

File: open_instruct/grpo_utils.py (hunk 4 above)
Comment (hamishivi) [reply open_instruct/grpo_utils.py]: discussed offline, this is already implemented here, i was just confused.
